src/idl_gen_general.cpp

```diff
@@ -874,14 +880,13 @@ class GeneralGenerator : public BaseGenerator {
     // Generate the __init method that sets the field in a pre-existing
     // accessor object. This is to allow object reuse.
     code += "  public void __init(int _i, ByteBuffer _bb) ";
-    code += "{ " + lang_.accessor_prefix + "bb_pos = _i; ";
-    code += lang_.accessor_prefix + "bb = _bb; ";
-    if (!struct_def.fixed && lang_.language == IDLOptions::kJava) {
-      code += lang_.accessor_prefix + "vtable_start = " + lang_.accessor_prefix + "bb_pos - ";
-      code += lang_.accessor_prefix + "bb." + FunctionStart('G') + "etInt(";
-      code += lang_.accessor_prefix + "bb_pos); " + lang_.accessor_prefix + "vtable_size = ";
-      code += lang_.accessor_prefix + "bb." + FunctionStart('G') + "etShort(";
-      code += lang_.accessor_prefix + "vtable_start); ";
+    code += "{ ";
+    if (lang_.language == IDLOptions::kCSharp) {
+      code += "__p = default(";
```

Comment:
using the `default` keyword in this way.. not seen it before in C#.. is this available in most versions of C#, including older versions like what e.g. Unity uses? Why is it needed beyond calling `__reset` ?
It's hard to Google for :)

Reply:
```C#
Table t;
private Struct __p;
```
These declarations reserve memory for value `t` and `__p` without initialization.
There are two ways to initialize struct by zeros without direct access to struct fields:
`t = new Struct()` or `t = default(Struct)`.
Both are assignments of the default value Struct() to `t`.
In both cases compiler calls `memset(&t, 0, sizeof(Struct))`.

The `default(T)` was introduced in C#2 (for generic types).
Flatbuffers generated C# code requires C#4 or higher.

Problem:
C# compiler expects that a program does initialization of all fields before the struct will be used (read/modify) or passed by value or reference.
In some (trivial) cases compiler can detect that struct is used without init and generate the error `CS0170: Use of possibly unassigned field 't'`.
The error generated only inside one method `TestTable.TestTable` if use proposed `__reset` method.
I assumed that `default (Table)` and `default (Struct)` would be useful in the generated code too.
Remove?

[C# structs](https://docs.microsoft.com/ru-ru/dotnet/csharp/programming-guide/classes-and-structs/using-structs)
[C# default value expression](https://docs.microsoft.com/lb-lu/dotnet/csharp/programming-guide/statements-expressions-operators/default-value-expressions)

Reply:
Ah, so `default` puts the variable in a state of a constructor/assignment not having been called, then it possibly errors. So it is the struct equivalent of `= null`, effectively. Since it is used on a previously initialized value, it would seem valid to remove it.

Frankly, even better would have been to not have a way to re-use handles in C# at all, since unlike Java they are cheap, so the user should instead always create a new one. Should we mark it as deprecated while we're at it?

Reply:
Issue-1.
I'm sure that understand you about `Since it is used on a previously initialized value, it would seem valid to remove it.`

For example, call sequence:
Step-1.
https://github.com/google/flatbuffers/blob/4eb3efc221d66ef02928d1b1860e097ab2e4ce16/tests/MyGame/Example/Monster.cs#L300
Step-2.
https://github.com/google/flatbuffers/blob/4eb3efc221d66ef02928d1b1860e097ab2e4ce16/tests/MyGame/Example/Monster.cs#L315
Step-3.
https://github.com/google/flatbuffers/blob/4eb3efc221d66ef02928d1b1860e097ab2e4ce16/tests/MyGame/Example/Monster.cs#L20
Step-4.
https://github.com/google/flatbuffers/blob/4eb3efc221d66ef02928d1b1860e097ab2e4ce16/tests/MyGame/Example/Monster.cs#L19


Step-1: We have Monster which has uninitialized struct `Table` __p.
>Unlike classes, structs can be instantiated without using the new operator. In such a case, there is no constructor call, which makes the allocation more efficient. However, the fields will remain unassigned and the object cannot be used until all of the fields are initialized. This includes the inability to get or set values through properties.

Step-4: Initialize all struct field before use. All is correct.

I propose to **hide** this initialization into private `__reset` method:
```
public void __init(int _i, ByteBuffer _bb) { __p = default(Table); __p.__reset(_i, _bb); }
```
Code must initialize all hidden fields of struct `__p` before calling the `__reset` method which hides this initialization. For this need call `default(Table)` or `new Table()`.
Right away It’s better to call the custom constructor: `new Table(_i, _bb)`.
But this call will cancel `struct` features.

Issue-2.
I don't know the reasons why `struct` was selected instead of `class`.
> Unless you need reference-type semantics, a small class may be more efficiently handled by the system if you declare it as a struct instead.

Array example: `testarrayofsortedstruct:[Ability]`.
But `pos` is scalar which returned as a reference to the interface (may be null).

An array of structs has an advantage over an array of references.
I don't know which is better. C# is not the primary language for me.

Reply:
> But this call will cancel struct features.

which? Replacing with a direct constructor (like in `__lookup_by_key` directly) would be preferable wherever possible.

> I don't know the reasons why struct was selected instead of class
'struct' in C# is way more efficient than class. It's stack vs heap allocation.

> Array example: testarrayofsortedstruct:[Ability]

But here we are talking about FlatBuffers `struct` which is subtly different from C# `struct` (though very similar).

> An array of structs has an advantage over an array of references.
I don't know which is better. C# is not the primary language for me.

I'd say in C# you want `struct` whenever possible.

Reply:
You are right. Direct call or ctor is preferable.
I checked it with SharpLab: [Sample .NET Core x64](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACXDKAVzAwYBVtgAbGAwDeNBmIb0GASwB2nYMAD6ABwi4A3KPGTZnCADMNNAJCSe/GAApdDRVICUwrePEKVahgF5bUzdRcuBrhePn4uAL7OEozEKLaKsLgwGNZyPo4i/gFibqrB3nZh2QxBIYVRkdSVNJJkDACyEDLMMFBOWWLKUFIAbtgYguYC8cp+UTppjc0DUAByMADulhlRLooqITKL3LwClgCs9kXZxADsIwB0uWrHDJUuE5xTLVAAIjD62Kx8KSsdAetlCEACYfL4/SxDGBHVbiQEXdaJZKWU4w/4Pc7w65Gf73cTjGLkJDSNIDZjLdrFBh9Nr4TbbZ4zZa3DEMfAXRmteZLRxwNkcpovd6fb6/W6VarUIA==)

Reply:
Oh wow, that's a great tool.. like godbolt.